live-build/misc/live-build-hooks/90-linux-migration-artifact.binary

```diff
@@ -0,0 +1,165 @@
+# to have "binary" as its top-level directory.
+#
+(
+	cd binary
+	find ./* -depth -print | cpio -ocB
+) >$DEPOT_DIRECTORY/os-root.cpio
+
+#
+# Copy all migration scripts including dx_prepare which is needed by
+# dx_unpack.
+#
+cp migration-scripts/* $DEPOT_DIRECTORY
+
+#
+# There may be a version.info file in the current directory already.
+# That version.info is Linux-specific and does not necessarily have
+# all the fields that the illumos upgrade process requires. Thus, we
+# create our own special version info directly in the depot directory
+# with all the expected fields plus a special variable that indicates
+# that this is a migration image, so dx_prepare knows how to tell the
+# difference between normal illumos upgrade and illumos-to-Linux
+# migrations.
+#
+KERNEL_VERSION=$(basename "$(find binary/boot/vmlinuz-* | sort -r | head -n 1)")
+{
+	echo "DLPX_DATE=$(date '+%Y.%m.%d.%H.%M.%S')"
+	echo "DLPX_OS_VERSION=$KERNEL_VERSION"
+	echo "DLPX_VERSION=$(date '+%Y.%m.%d.%H')"
+	echo "DLPX_MIN_VERSION=5.3.0.0"
+	echo "MIGRATION_IMAGE=y"
+} >$DEPOT_DIRECTORY/version.info
+
+#
+# Generate hashes file.
+#
+(
+	cd binary
+	find ./* -type f -print0 | xargs -0 sha256sum
+) >$DEPOT_DIRECTORY/os-root.hashes || die "generating os-root.hashes failed"
+(
+	cd $DEPOT_DIRECTORY
+	sha256sum ./* >hashes
+)
+
+#
+# DLPX_SIGNING_LOGIN is used for signing our migration image.
+# The signature is later unpacked and verified on the VM
+# performing the migration. If DLPX_SIGNING_LOGIN has not been
+# specified by the user when running this hook we skip generating
+# the signature for this image.
+#
+# Note that DLPX_KEY_URL should also be specified for this to
+# work, as the default is really a placeholder.
+#
+if [ "$DLPX_SIGNING_LOGIN" != "delphix" ]; then
+	#
+	# Generate depot/hashes.sig[.<release>]
+	#
+	# Assumption: we always migrate from version 5.3
+	#
+	UPGRADE_VERSION="5.3"
+	DLPX_SIGNING_URL="$DLPX_KEY_URL/upgrade/keyVersion/$UPGRADE_VERSION/sign"
+
+	hashes_data=$(base64 -w 0 $DEPOT_DIRECTORY/hashes)
+	json_payload="{\"data\": \"${hashes_data}\"}"
+	request_url="$DLPX_SIGNING_URL"
+
+	#
+	# Compute signature
+	#
+	result=$(
+		set -o pipefail
+		echo "$json_payload" | curl -s -f -H \
+			"Content-Type: application/json" \
+			-u "$DLPX_SIGNING_LOGIN" "$request_url" -d @-
+	) || die "Failure signing hashes with $request_url"
+
+	#
+	# Extract signature from JSON response, ensure it is not empty
+	# and save it.
+	#
+	signature=$(echo "$result" | grep signature |
+		sed 's/.*"signature": "\([^"]*\)".*/\1/')
+	[ -n "$signature" ] || die "Empty signature obtained with $request_url"
+	echo "$signature" | base64 -d >$DEPOT_DIRECTORY/hashes.sig.$UPGRADE_VERSION
+fi
+
+#
+# Tape-ARchive & Compress Zee File!
+#
+# We enable verbose output in the hope that all the files that are part of
+# the artifact are logged somewhere which could aid debugging in the future.
+#
+tar -cvf - $DEPOT_DIRECTORY | pigz >"$MIGRATION_ARTIFACT"
```

Comment:
We probably want to use `pigz` for the upgrade image too; after this lands, can you open a bug so we can circle back to this for the upgrade image?

Reply:
https://github.com/delphix/appliance-build/issues/157